lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +796,585 @@ def __str__(self):
+        if self.ref_type == 'Branch':
+            return True
+        assert self.ref_type in ['Tag', 'Commit'], self
+        return False
+
+    def _repo_info(self):
+        # type: () -> str
+        if self.ref_type == 'Commit':
+            return 'at commit {0}'.format(self._ref)
+        if self.ref_type == 'Tag':
+            return 'at tag {0}'.format(self._ref)
+        assert self.ref_type == 'Branch', self
+        return 'on branch {0}'.format(self._ref)
+
+    def repo_info_for_reference(self):
+        # type: () -> str
+        if self.should_validate_matches_hash:
+            return '{} (at commit {})'.format(
+                self._repo_info(), self.should_validate_matches_hash)
+        return self._repo_info()
+
+
+class GitCommit(GitRef):
+    """A git commit."""
+
+    ref_type = 'Commit'        # type: ClassVar[str]
+
+    # Matches any 7-40 character hexadecimal string.
+    _commit_rx = re.compile(r'^[0-9a-f]{7,40}$', flags=re.IGNORECASE)
+
+    @property
+    def should_validate_matches_hash(self):
+        # type: () -> Optional[str]
+        """There is no need to verify that a git commit hash matches itself."""
+        return None
+
+    def __init__(self, ref):
+        # type: (str) -> None
+        if not isinstance(ref, six.string_types):
+            # The exception message from `re.match()` does not print the value when
+            # a non-string argument is provided, so we special-case that check here.
+            raise InvalidGitRef('git reference was not a string: {0}'.format(ref))
+        if not self._commit_rx.match(ref):
+            raise InvalidGitRef(dedent("""\
+            Spack requires that references to individual git commits be specified via
+            a 7-40 character hexadecimal string, but received {0!r} instead.
+
+            A valid commit hash case-insensitively matches the regular expression '{1}'.
+            7 hex characters is the size printed out by `git log --format='%h'`, while
+            40 hex characters is the size printed out by `git log --format='%H'`.
+            """.format(ref, self._commit_rx.pattern)))
+        super(GitCommit, self).__init__(ref)
+GitRef.Commit = GitCommit       # noqa: E305
+
+
+class GitTag(GitRef):
+    """A git tag."""
+
+    ref_type = 'Tag'        # type: ClassVar[str]
+
+    @property
+    def version(self):
+        # type: () -> spack.version.Version
+        return spack.version.Version(self._ref)
+
+    @property
+    def should_validate_matches_hash(self):
+        # type: () -> Optional[str]
+        return self._commit
+
+    def __init__(self, ref, commit=None):
+        # type: (str, Optional[str]) -> None
+        super(GitTag, self).__init__(ref)
+        self._commit = commit
+
+    def __repr__(self):
+        return 'GitTag(ref_type={!r}, ref={!r}, commit={!r})'.format(
+            self.ref_type, self._ref, self._commit,
+        )
+
+GitRef.Tag = GitTag             # noqa: E305
```

Comment:
Just a thought - and I'm sort of torn. I like this way because you can see the class added right after, but another idea is to group them together so it's more tidy. I guess I don't feel strongly either way but wanted to mention it in case someone else feels strongly.

Reply:
Was also torn over the same thing!